doc: improve fetchers overview, deduplicate readme content, follow doc conventions

DanielSidhion opened the pull request:
## Description of changes

This PR starts my work to update and improve the chapter on fetchers in the Nixpkgs manual. It adds more explanations on the overview of the chapter, moves content from the contributor README into the manual, and deduplicates information.

**NOTE:** this PR started as a change of the fetchers overview AND updating the documentation for `fetchurl`. After a lot of reviewing, we decided to split this PR into two. The PR concerns the description above. The second PR is #300429 and concerns only with updating the documentation for `fetchurl`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

DanielSidhion committed:
doc: improve fetchers overview, deduplicate readme content

lolbinarycat commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -23,66 +26,444 @@ The following table shows an overview of the differences:
+
+### File naming {#sec-pkgs-fetchers-fetchurl-file-naming}
+
+The name of the file within the Nix store is controlled by the hash given to `fetchurl`, and also the `name` (or `pname` and `version`) values.
```
> The name of the file within the Nix store

I believe the official term is "derivation name"

DanielSidhion replied:
The Nix manual uses "derivation name" for the symbolic part of the name (which in the case of `fetchurl` would be determined by the `name` or `pname` + `version` attributes). I wanted to refer to the file name in this case, but the naming gets very confusing very fast (see below), so I prefer to stick to "name of the file within the Nix store" for now. At some point I want to attach a glossary to the Nixpkgs manual to make it easier to refer to names and their definitions, and at that point I hope I can start clarifying definitions in other manuals as well.

Why it gets very confusing very fast:
- In theory what I want to refer should be called the store path, but the [Nix manual glossary](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) has a definition for store path which is not coherent with the [store path section](https://nixos.org/manual/nix/stable/store/store-path) in the same manual.
- Calling it a "file within the Nix store" is also not entirely true because there can be Nix stores that aren't part of the filesystem, but I couldn't find any better term that is easily understandable for the main audience of the manual.

Given that most people reading the manual will think of the Nix store as the one in their local filesystem, I chose to use that. The people who are aware of other store types and its quirks should be able to understand that the "file" part doesn't matter and this section is talking about the name of the object in the store more generally.

lolbinarycat replied:
ah, here you are referring to hash+name, but i believe earlier you used the same wording to describe just the derivation name?

DanielSidhion replied:
It was later when describing the `name`, `pname`, and `version` attributes. I just pushed a change to be more specific on what kind of "name" I'm talking about, while trying as much as possible to avoid more confusion.

fricklerhandwerk replied:
```suggestion
The resulting store path is determined by the hash given to `fetchurl`, and also the `name` (or `pname` and `version`) values.
```

And maybe use the Nix manual link from the line below, then we don't need an entire extra sentence.

DanielSidhion replied:
I split this PR into two. The `fetchurl` changes are in #300429 where I applied this suggestion.

lolbinarycat reviewed (commented):
wow that's a lot of documentation!

DanielSidhion committed:
Improve caveat explanation and some fetchurl content

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -23,66 +26,444 @@ The following table shows an overview of the differences:
-For example, consider the following fetcher:
+Because Nixpkgs fetchers are [fixed-output derivations](https://nixos.org/manual/nix/stable/glossary#gloss-fixed-output-derivation), they need you to specify a hash (this is what turns them into a fixed-output derivation).
+This is usually done through a `hash` attribute.
+However, **this hash is not a hash of the contents that are downloaded**, it is a hash that corresponds to the derivation output itself, and is calculated by Nix.
```
Thank you! Indeed, after reading this section again I wasn't happy with how things were explained, and tried to improve it.

fricklerhandwerk committed:
reorder statements

Co-authored-by: Dominic Mills-Howell <dominic.millz27@gmail.com>

fricklerhandwerk committed:
refine wording

Co-authored-by: Dominic Mills-Howell <dominic.millz27@gmail.com>

fricklerhandwerk commented:
Started reviewing in docs team meeting today.

This is a rabbit hole, because there is overlapping [documentation on updating FOD hashes](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#sources) that ended up in contributor documentation, where it arguably shouldn't be. There are also repetitions of the same mantra in [various other places](https://github.com/search?q=fakeHash+repo%3Anixos%2Fnixpkgs+path%3A*.md&type=Code&ref=advsearch&l=&l=). Ideally we'd merge all of that in the manual's fetcher section.

I'll push a few commits to improve the wording on the new stuff, but recommend splitting out the reference documentation in a different PR to get some of that work merged already.

fricklerhandwerk committed:
move out consumer docs on source fetching

update wording in the contributors docs

Co-authored-by: Dominic Mills-Howell <dominic.millz27@gmail.com>

fricklerhandwerk committed:
move note on mirror URLs to the relevant section

this may be better suited for the `fetchurl` reference, but I'd prefer
just rendering that information into the manual. for now, because
- contributor documentation encourages mirrors
- we can expect contributors to dig into the source
- linking source files is trivial in in-code documentation
we leave it there.

Co-authored-by: Dominic Mills-Howell <dominic.millz27@gmail.com>

fricklerhandwerk committed:
move instructions for updating hashes to the manual

also improve the language a bit, and add links

Co-authored-by: Dominic Mills-Howell <dominic.millz27@gmail.com>

fricklerhandwerk commented:
@DMills27 and me went all the way and closed the rabbit hole.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-03-28-documentation-team-meeting-notes-115/42329/1

fricklerhandwerk commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -1,88 +1,540 @@
-This significantly reduces the time needed to evaluate the entirety of Nixpkgs, and allows [Hydra](https://nixos.org/hydra) to retain and re-distribute sources used by Nixpkgs in the [public binary cache](https://cache.nixos.org).
-For these reasons, built-in fetchers are not allowed in Nixpkgs source code.
+This significantly reduces the time needed to evaluate code in Nixpkgs, and allows [Hydra](https://nixos.org/hydra) to retain and re-distribute sources used by Nixpkgs in the [public binary cache](https://cache.nixos.org).
+
```
```suggestion
```

fricklerhandwerk commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -1,88 +1,540 @@
+For these reasons, the Nix built-in fetchers are not allowed in Nixpkgs code.
 
-The following table shows an overview of the differences:
+The following table summarizes the differences:
```
```suggestion
The following table summarises the differences:
```

fricklerhandwerk commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -1,88 +1,540 @@
+
+It uses {manpage}`curl(1)` internally, and allows its behaviour to be modified by specifying a few attributes in the argument to `fetchurl` (see the documentation for attributes `curlOpts`, `curlOptsList`, and `netrcPhase`).
+
+### File naming {#sec-pkgs-fetchers-fetchurl-file-naming}
```
```suggestion
```

Don't think this is needed. The section is not that long.

DanielSidhion replied:
Thanks! I split this PR into two, and I applied this suggestion in #300429 which is the PR with the fetchurl changes.

fricklerhandwerk committed:
british english

fricklerhandwerk committed:
remove stray empty line

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
-Because Nixpkgs fetchers are [fixed-output derivations](https://nixos.org/manual/nix/stable/glossary#gloss-fixed-output-derivation), they need you to specify a hash (this is what turns them into a fixed-output derivation).
-This is usually done through a `hash` attribute, which **can be different from the hash of the contents that are downloaded**.
-On top of that, Nix checks if there is already an object in the Nix store with the same hash, and reuses that object if it exists.
+Because Nixpkgs fetchers are fixed-output derivations, an [output hash](https://nixos.org/manual/nix/stable/language/advanced-attributes#adv-attr-outputHash) has to be specified.
```
```suggestion
Because Nixpkgs fetchers are fixed-output derivations, an [output hash](https://nixos.org/manual/nix/stable/language/advanced-attributes#adv-attr-outputHash) has to be specified, usually indirectly through a `hash` attribute.
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
-This is usually done through a `hash` attribute, which **can be different from the hash of the contents that are downloaded**.
-On top of that, Nix checks if there is already an object in the Nix store with the same hash, and reuses that object if it exists.
+Because Nixpkgs fetchers are fixed-output derivations, an [output hash](https://nixos.org/manual/nix/stable/language/advanced-attributes#adv-attr-outputHash) has to be specified.
+This hash refers to the derivation output, not the remote source itself!
```
```suggestion
This hash refers to the derivation output, which can be different from the remote source itself!
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
+
+- When changing any fetcher parameters, always update the output hash.
+  Use one of the methods from [](#sec-pkgs-fetchers-updating-source-hashes).
+  Otherwise, existing store objects that match the output hash will be re-used rather than fetching new content.
+
+A similar problem arises while testing changes to a fetcher's implementation.
+If the output of the derivation already exists in the Nix store, test failures can go undetected.
+The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.
```
```suggestion
  Otherwise, existing store objects that match the output hash will be re-used rather than fetching new content.

  :::{.note}
  A similar problem arises while testing changes to a fetcher's implementation.
  If the output of the derivation already exists in the Nix store, test failures can go undetected.
  The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.
  :::
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
+
+## Updating source hashes {#sec-pkgs-fetchers-updating-source-hashes}
+
+There are several ways to obtain a remote source's hash.
```
```suggestion
There are several ways to obtain the hash corresponding to a remote source.
Unless you understand how the fetcher you're using calculates the hash from the downloaded contents, you should use [the fake hash method](#sec-pkgs-fetchers-updating-source-hashes-fakehash-method).
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
+   Don't use it when upstream provides `md5`, compute `sha256` instead.
 
-For example, consider the following recipe that only uses a fetcher:
+   A little nuance is that `nix-prefetch-*` tools produce hashes encoded with the custom Nix `base32` variant, but upstream usually provides hexadecimal (`base16`) encoding.
```
```suggestion
   A little nuance is that `nix-prefetch-*` tools produce hashes with the `nix32` encoding (a Nix-specific base32 adaptation), but upstream usually provides hexadecimal (`base16`) encoding.
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
+4. Extract the hash from a local source archive with `sha256sum`.
+   Use `nix-prefetch-url file:///path/to/archive` if you want the custom Nix `base32` hash.
+
+5. Trust on first use: set the hash to one of
```
Every other method is also trust-on-first-use depending on how you see it. This is the only method that someone can follow without having to learn beforehand how exactly a fetcher calculates the hash (e.g. `fetchzip` uses recursive hash, so blindly using `nix-prefetch-url` or `sha256sum` won't give the correct hash).

```suggestion
5. []{#sec-pkgs-fetchers-updating-source-hashes-fakehash-method} In your package recipe, set the hash to one of
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
+   - `lib.fakeSha256`
+   - `lib.fakeSha512`
+
+   in the package recipe, attempt to build, and extract the correct hash from error messages.
```
```suggestion
   Once you attempt to build, error messages with the calculated hashes will show up, which you then replace in the recipe.
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
+   You must use one of these four fake hashes and not some arbitrarily-chosen hash.
+   See [][#sec-pkgs-fetchers-secure-hashes] for details.
+   :::
+
+   This is last resort method when reconstructing source URL is non-trivial and `nix-prefetch-url -A` isn’t applicable.
```
```suggestion
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
+
+:::{.example #ex-fetchers-update-fod-hash}
+
+# Update source hash with "trust on first use"
```
```suggestion
# Update source hash with the fake hash method
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -58,9 +116,7 @@ fetchurl {
-**This will reuse the old contents**.
-Remember to invalidate the hash argument.
-You can do this by setting the `hash` attribute to an empty string.
+**This will produce the same output as before!**.
```
```suggestion
**This will produce the same output as before!**.
To determine the new hash, set it to an empty value (it's a fake hash, hence the name of the method):
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
````diff
@@ -70,28 +126,32 @@ fetchurl {
 ```shell
 $ nix-build
-(some output removed for clarity)
+...
````
I prefer to be verbose in this case: most people who are familiar with CLIs will understand both "..." and "(some output removed for clarity)", but there is a small amount of less experienced folks who will get a wall of text of output and think something went wrong and never actually read the output.

Using "(some output removed for clarity)" gives them indirect encouragement to read through the wall of text until they find the text they should be reading.

```suggestion
(some output removed for clarity)
```

fricklerhandwerk replied:
Hm, I'd rather stick to the widely used convention and (at some point in the future) add a page on how to read examples.

DanielSidhion replied:
I decided to defer on this one since there is already a bunch of content (from previous PRs) where I used `(some output removed for clarity)` or `(output removed for clarity)`, there are also places in the manual where `...` is used, and other places where nothing is used.

I'd like to make this part of the linting tool for docs content, which is in [the roadmap](https://discourse.nixos.org/t/a-roadmap-for-the-documentation-ecosystem/42328) for docs, so let's make this decision in the issue for the linting tool itself. And then we can apply it to all existing content to make things consistent at once.

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -70,28 +126,32 @@ fetchurl {
 
-Remember to set the `hash` attribute to an empty string whenever you update fetcher parameters and expect the content to change!
-:::
+Let's say Man-in-the-Middle (MITM) sits close to your network.
+Then instead of fetching source you can fetch malware, and instead of the actual source hash, you get the hash of malware.
```
```suggestion
It's always a good idea to avoid Man-in-the-Middle (MITM) attacks when downloading source contents.
Otherwise, you could unknowingly download malware instead of the intended source, and instead of the actual source hash, you'll end up using the hash of malware.
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -70,28 +126,32 @@ fetchurl {
+
+- Upstream hashes should be obtained via a secure protocol.
+
+- `https://` URLs are secure when using `nix-prefetch-*` or for upstream hashes.
```
```suggestion
- `https://` URLs give you more protections when using `nix-prefetch-*` or for upstream hashes.
```

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -70,28 +126,32 @@ fetchurl {
+
+- `https://` URLs are secure when using `nix-prefetch-*` or for upstream hashes.
+
+- `https://` URLs are secure relying on "trust on first use" *only if* you use one of the listed fake hashes.
+  If you use any other hash, `fetchurl` will pass `--insecure` to `curl` and may then degrade to HTTP in case of TLS certificate expiration.
```
Mentioning `fetchurl` details will feel confusing here. If this is new information for anyone reading this, chances are very low that they will understand what "`fetchurl` will pass `--insecure` to `curl`" and what "TLS certificate expiration" means.

I'm adapting this to be most understandable to the audience which will find this section relevant.

```suggestion
- `https://` URLs are secure when using the [fake hash method](#sec-pkgs-fetchers-updating-source-hashes-fakehash-method) *only if* you use one of the listed fake hashes.
  If you use any other hash, the download will be exposed to MITM attacks even if you use HTTPS URLs.
```

fricklerhandwerk replied:
No, security-related information should be concrete. This smudges away the facts. We can be a bit more clear and say

> will pass `--insecure` to the underlying call to` curl`

DanielSidhion replied:
Added this information, I think it covers all audiences now.

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -17,28 +13,90 @@ The differences are explained in this list:
 | `builtins.fetch*` | evaluation time | store path | `/nix/store`, `~/.cache/nix` | `tarball-ttl` expires, cache miss in `~/.cache/nix`, output store object not in local store |
 | `pkgs.fetch*` | build time | derivation | `/nix/store`, substituters | output store object not available |
 
+:::{.note}
```
```suggestion
:::{.tip}
```

DanielSidhion committed:
Apply suggestions from code review and fix some typos

DanielSidhion commented:
@fricklerhandwerk @lolbinarycat: I followed @fricklerhandwerk's suggestion and split this PR into two. Did a bit of git history rewriting to get only the relevant changes in this PR, and #300429 has the fetchurl changes.

@fricklerhandwerk - I made some changes and reworded some stuff on top of your changes, please take a look at the comments [from my review](https://github.com/NixOS/nixpkgs/pull/297654#pullrequestreview-1970112030) to understand the reasoning. Requesting your review and @lolbinarycat's and I'm hoping we can agree on a merge soon :blush: 

Thank you all for the reviews, I feel this has turned out much better with your involvement!

fricklerhandwerk commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -1,66 +1,159 @@
-**This will reuse the old contents**.
-Remember to invalidate the hash argument, in this case by setting the `hash` attribute to an empty string.
+**This will produce the same output as before!**
+To determine the new hash, set it to an empty value (it's a fake hash, hence the name of the method):
```
```suggestion
Set the hash to an empty string:
```

fricklerhandwerk reviewed (commented):
Just a few small things. The other changes are clear improvements, thanks a lot!

lolbinarycat commented on pkgs/README.md:
```diff
@@ -406,13 +402,16 @@ Import From Derivation can be worked around in some cases by committing generate
+Prefer protocols that support proxies.
 
-There are multiple ways to fetch a package source in nixpkgs. The general guideline is that you should package reproducible sources with a high degree of availability. Right now there is only one fetcher which has mirroring support and that is `fetchurl`. Note that you should also prefer protocols which have a corresponding proxy environment variable.
+Currently only [`fetchurl`](https://nixos.org/manual/nixpkgs/unstable/#fetchurl) has mirroring support.
+A list of schemes for `mirror://` URLs can be found in [`pkgs/build-support/fetchurl/mirrors.nix`](build-support/fetchurl/mirrors.nix).
```
did you check `fetchFrom*`? none of those use mirror urls?

DanielSidhion replied:
I updated the sentences to explain that `fetchurl` has **direct** mirroring support, and other fetchers may support mirroring if they end up using `fetchurl`.

lolbinarycat reviewed (commented):
look pretty good but i think there are some slight improvements that can be made.

DanielSidhion committed:
Apply suggestions from code review

Co-authored-by: lolbinarycat <dogedoge61+github@gmail.com>
Co-authored-by: Valentin Gagarin <valentin.gagarin@tweag.io>

DanielSidhion commented on doc/build-helpers/fetchers.chapter.md:
```diff
@@ -1,66 +1,159 @@
 | `builtins.fetch*` | evaluation time | store path | `/nix/store`, `~/.cache/nix` | `tarball-ttl` expires, cache miss in `~/.cache/nix`, output store object not in local store |
 | `pkgs.fetch*` | build time | derivation | `/nix/store`, substituters | output store object not available |
 
+:::{.tip}
+`pkgs.fetchFrom*` helpers retrieve _snapshots_ of version-controlled sources, as opposed to the entire version history, which is more efficient.
+:::
+
```
I'll include `fetchgit` in the note. Regarding mirroring: I'll come back to this and update it with mirroring information as I make my way through documenting all the fetchers, because then I'll see which fetcher supports mirroring and which doesn't, whether there are any differences in support, and so on.

It's likely this tip will go away once all fetchers are better documented, because by then I'll have a better idea of how to organise all this fetcher-agnostic content as well.

DanielSidhion committed:
Add more clarity on text, reorganise source hash methods

fricklerhandwerk reviewed (approved):
This has probably seen enough churn now. Let's fix forward if rough edges remain.
